### PR TITLE
Multi-cq

### PR DESCRIPTION
**Pull request by DrJessop**
This PR will enable us to have multiple hardware CQs on a single device. Currently, the CommandQueue class is one-to-one with our concept of hardware command queues.

As a next step to expose >2 software command queues, we will rename CommandQueue to HardwareCommandQueue and have a higher level class called CommandQueue which will serialize it's commands into a particular HardwareCommandQueue's 'stream'. By doing it this way, we don't need to worry about mutexing all of the CommandQueueWriter APIs.

The amount of hardware resources a user would like currently needs to be known when they initialize a device. Based on this number of resources, we will partition the hugepage into that number of chunks. Given so far we are supporting at most 2 sets of resources, that means that each hardware command queue will have access to 512MB.

When we expose software command queues, as mentioned, they will serialize, and based on that serialization, the data will be written to a particular hardware CQ's system resources in order.

### CODE REVIEW

**Review by boris-drazic — APPROVED** [review]

**Review by farbabi — APPROVED** [review]

File: tt_metal/impl/device/device.hpp [hunk 1]
```diff
@@ -129,8 +131,10 @@ class Device {
     // Set of logical storage only core coordinates
     const std::set<CoreCoord> &storage_only_cores() const { return this->storage_only_cores_; }
 
+    const std::set<CoreCoord> &producer_cores() const { return this->producer_cores_; }
```
[the remark this answers is not in the document]
Comment (DrJessop) [reply tt_metal/impl/device/device.hpp:134]: No, one pair of producer and consumer == 1 hw cq. So num_hw_cqs = num producer cores = num consumer cores

File: tt_metal/host_api.hpp [hunk 2]
```diff
@@ -56,11 +56,11 @@ size_t GetNumPCIeDevices();
...
  *
- * | Argument  | Description                | Type            | Valid Range                       | Required |
- * |-----------|----------------------------|-----------------|-----------------------------------|----------|
- * | device_id | ID of the device to target | chip_id_t (int) | 0 to (GetNumAvailableDevices - 1) | Yes      |
+ * | Argument   | Description                | Type            | Valid Range                       | Required |
+ * |------------|----------------------------|-----------------|-----------------------------------|----------|
+ * | device_id  | ID of the device to target| chip_id_t (int) | 0 to (GetNumAvailableDevices - 1) | Yes      |
  * */
-Device *CreateDevice(chip_id_t device_id, const std::vector<uint32_t>& l1_bank_remap = {});
+Device *CreateDevice(chip_id_t device_id, const uint8_t num_hw_cqs = 1, const std::vector<uint32_t>& l1_bank_remap = {});
```
[the remark this answers is not in the document]
Comment (DrJessop) [reply tt_metal/host_api.hpp:63]: the yaml contains all the possible configurations. we need to specify which config we want since there is compile time information inferred based on the num of hw cqs

[the remark this answers is not in the document]
Comment (TT-billteng) [reply tt_metal/host_api.hpp:63]: also will user be able to put any number of HW CQs as a parameter (3, 4, 10, 100)? If not, this level of configurability shouldn't be exposed

[the remark this answers is not in the document]
Comment (TT-billteng) [reply tt_metal/host_api.hpp:63]: is it possible to use only the 2 HW CQ yaml config from here on out? You'd need to be backwards compatiable with user requesting only 1 queue. This would remove the need for configuring it here and also remove need for a separate GTest test suite

File: tt_metal/hw/inc/mod_div_lib.h [hunk 3]
```diff
@@ -42,6 +47,8 @@ inline __attribute__((always_inline)) uint32_t udivsi3_const_divisor(uint32_t n)
     } else if constexpr (d == 94) {
         // fast divide for 94 divisor. Handles Banked L1 address generation for E75
         return fast_udiv_94(n);
+    } else if constexpr (d == 124) {
```
Comment (TT-billteng) [inline tt_metal/hw/inc/mod_div_lib.h:50]: how is this change related to this PR?

File: tests/tt_metal/tt_metal/perf_microbenchmark/noc/test_noc_unicast_vs_multicast_to_single_core_latency.cpp [hunk 4]
```diff
@@ -11,14 +11,14 @@
...
 void measure_latency(string kernel_name) {
     const int device_id = 0;
     tt_metal::Device *device = tt_metal::CreateDevice(device_id);
 
-    auto dispatch_cores = device->dispatch_cores().begin();
-    CoreCoord producer_logical_core = *dispatch_cores++;
-    CoreCoord consumer_logical_core = *dispatch_cores;
+    // auto dispatch_cores = device->dispatch_cores().begin();
+    CoreCoord producer_logical_core = {0, 0};//*dispatch_cores++;
+    CoreCoord consumer_logical_core = {0, 0}; //*dispatch_cores;
```
Comment (abhullar-tt) [inline tests/tt_metal/tt_metal/perf_microbenchmark/noc/test_noc_unicast_vs_multicast_to_single_core_latency.cpp:21]: I think this needs to be updated

File: tests/tt_metal/tt_metal/unit_tests/basic/device.cpp [hunk 5]
```diff
@@ -352,7 +352,6 @@ TEST_F(DeviceFixture, TestDeviceToHostMemChannelAssignment) {
 
     for (const auto& [mmio_dev_id, device_group] : mmio_device_to_device_group) {
         EXPECT_EQ(tt::Cluster::instance().get_num_host_channels(mmio_dev_id), device_group.size());
-        EXPECT_EQ(tt::Cluster::instance().get_assigned_channel_for_device(mmio_dev_id), 0);
```
Comment (abhullar-tt) [inline tests/tt_metal/tt_metal/unit_tests/basic/device.cpp:355]: there was no real reason for mmio device to be given channel 0 hugepage but curious if this was changed?

File: tests/tt_metal/tt_metal/unit_tests_fast_dispatch/command_queue/test_EnqueueWriteBuffer_and_EnqueueReadBuffer.cpp [hunk 6]
```diff
@@ -329,8 +327,8 @@ TEST_F(CommandQueueFixture, TestWrapHostHugepageOnEnqueueReadBuffer) {
     uint32_t num_pages = buffer_size / page_size;
 
     TestBufferConfig buf_config = {.num_pages = num_pages, .page_size = page_size, .buftype = BufferType::DRAM};
-
-    EXPECT_TRUE(local_test_functions::test_EnqueueWrap_on_EnqueueReadBuffer(this->device_, tt::tt_metal::detail::GetCommandQueue(device_), buf_config));
+    CommandQueue a(this->device_, 0);
```
Comment (abhullar-tt) [inline tests/tt_metal/tt_metal/unit_tests_fast_dispatch/command_queue/test_EnqueueWriteBuffer_and_EnqueueReadBuffer.cpp:330]: any specific reason to create CQ explicitly here?

File: tests/tt_metal/tt_metal/unit_tests_fast_dispatch/common/command_queue_fixture.hpp [hunk 7]
```diff
@@ -62,3 +62,30 @@ class MultiCommandQueueFixture : public ::testing::Test {
     tt::ARCH arch_;
     size_t num_devices_;
 };
+
+class CommandQueueMultiDeviceFixture : public ::testing::Test {
```
Comment (abhullar-tt) [inline tests/tt_metal/tt_metal/unit_tests_fast_dispatch/common/command_queue_fixture.hpp:66]: whats the difference between this and `MultiCommandQueueFixture` and also is it just me or is `num_devices_` undefined?

File: tests/tt_metal/tt_metal/perf_microbenchmark/noc/test_noc_unicast_vs_multicast_to_single_core_latency.cpp [hunk 8]
```diff
@@ -11,14 +11,14 @@
...
+//
 void measure_latency(string kernel_name) {
     const int device_id = 0;
     tt_metal::Device *device = tt_metal::CreateDevice(device_id);
 
-    auto dispatch_cores = device->dispatch_cores().begin();
-    CoreCoord producer_logical_core = *dispatch_cores++;
-    CoreCoord consumer_logical_core = *dispatch_cores;
+    // auto dispatch_cores = device->dispatch_cores().begin();
+    CoreCoord producer_logical_core = {0, 0};//*dispatch_cores++;
```
Comment (TT-billteng) [inline tests/tt_metal/tt_metal/perf_microbenchmark/noc/test_noc_unicast_vs_multicast_to_single_core_latency.cpp:20]: why is the core here (0,0)? The yaml specifies different cores as the producer/consumer pair.

File: tt_metal/detail/tt_metal.hpp [hunk 9]
```diff
@@ -284,9 +284,7 @@ namespace tt::tt_metal{
             static std::mutex cq_creation_mutex;
             {
                 std::lock_guard<std::mutex> lock(cq_creation_mutex);
-                if (not command_queues[id] or (command_queues[id] and command_queues[id]->device != device)) {
-                    command_queues[device->id()] = std::make_unique<CommandQueue>(device);
-                }
+                command_queues[device->id()] = std::make_unique<CommandQueue>(device, id);
```
Comment (abhullar-tt) [inline tt_metal/detail/tt_metal.hpp:287]: could you add a comment to the GetCommandQueue function explaining why a new CQ instance is created each time this API is called

File: tt_metal/impl/dispatch/command_queue.hpp [hunk 10]
```diff
@@ -186,17 +191,37 @@ namespace detail{
 
 class CommandQueue {
    public:
-    CommandQueue(Device* device);
+    CommandQueue(Device* device, uint32_t command_queue_channel);
```
Comment (TT-billteng) [inline tt_metal/impl/dispatch/command_queue.hpp:194]: Can command_queue_channel be inferred somehow? This is something user will be creating more often going forward

File: tt_metal/detail/tt_metal.hpp [hunk 11]
```diff
@@ -384,7 +381,7 @@ namespace tt::tt_metal{
                 soc_d.get_physical_ethernet_cores(),
                 soc_d.grid_size,
                 harvested_rows,
-                {device->worker_core_from_logical_core(consumer_logical_core)});
+                device->worker_core_from_logical_core(enqueue_program_dispatch_core));
```
Comment (abhullar-tt) [inline tt_metal/detail/tt_metal.hpp:384]: do we only supply one consumer core because currently only one cq (the first cq) will be servicing enqueue program commands?

File: tt_metal/detail/tt_metal.hpp [hunk 12]
```diff
@@ -445,77 +442,92 @@ namespace tt::tt_metal{
...
 
-            // TODO(pkeller): Should use detail::LaunchProgram once we have a mechanism to avoid running all RISCs
-            tt::llrt::write_launch_msg_to_core(device->id(), producer_physical_core, &msg);
-            tt::llrt::write_launch_msg_to_core(device->id(), consumer_physical_core, &msg);
+            const uint32_t cq_size = tt::Cluster::instance().get_host_channel_size(device->id(), hugepage_channel) / device->producer_cores().size();
+            TT_ASSERT(device->producer_cores().size() == device->consumer_cores().size(), "There must be the same number of producers as there are consumers");
+            TT_ASSERT(device->producer_cores().size() > 0, "There must be at least 1 producer/consumer core");
+            TT_ASSERT(device->producer_cores().size() < 3, "There can be at most 2 hardware command queues on a given device");
+
+            uint8_t cq_channel = 0;
```
Comment (abhullar-tt) [inline tt_metal/detail/tt_metal.hpp:455]: I think cq_channel here is just cq_id?

File: tt_metal/impl/dispatch/command_queue.cpp [hunk 13]
```diff
@@ -960,42 +945,58 @@ void CommandQueue::enqueue_program(Program& program, bool blocking) {
...
                 this->device, program_data_size_in_bytes, DeviceCommand::PROGRAM_PAGE_SIZE, BufferType::DRAM));
 
-        this->enqueue_write_buffer(*this->program_to_buffer.at(program_id), program_pages.data(), blocking);
-        this->program_to_dev_map.emplace(program_id, std::move(program_to_device_map));
+        this->enqueue_write_buffer(*program_to_buffer.at(program_id), program_pages.data(), blocking);
+
+        map<uint64_t, ProgramMap>& program_to_dev_map = this->program_to_dev_map(device->id());
+        program_to_dev_map.emplace(program_id, std::move(program_to_device_map));
+
+        const char *READ_BACK_PROGRAMS = std::getenv("TT_METAL_READ_BACK_PROGRAMS");
```
Comment (abhullar-tt) [inline tt_metal/impl/dispatch/command_queue.cpp:969]: did we always have this env var?

File: tt_metal/impl/dispatch/kernels/command_queue_consumer.cpp [hunk 14]
```diff
@@ -53,19 +57,22 @@ void kernel_main() {
...
+            wait_for_program_completion(num_workers);
         } else {
             command_ptr = reinterpret_cast<volatile tt_l1_ptr uint32_t*>(buffer_transfer_start_addr);
-            write_buffers(command_ptr, command_issue_region_size, num_buffer_transfers,  sharded_buffer_num_cores, consumer_cb_size, consumer_cb_num_pages, producer_noc_encoding, producer_consumer_transfer_num_pages, db_buf_switch);
+            write_buffers(command_ptr, completion_queue_start_addr, num_buffer_transfers,  sharded_buffer_num_cores, consumer_cb_size, consumer_cb_num_pages, producer_noc_encoding, producer_consumer_transfer_num_pages, db_buf_switch);
         }
 
         if (finish) {
-            notify_host_complete();
+            // DPRINT << "GOT FINISH" << ENDL();
```
Comment (abhullar-tt) [inline tt_metal/impl/dispatch/kernels/command_queue_consumer.cpp:74]: debug print

File: tt_metal/impl/dispatch/kernels/command_queue_consumer.cpp [hunk 15]
```diff
@@ -53,19 +57,22 @@ void kernel_main() {
         uint32_t wrap = command_ptr[DeviceCommand::wrap_idx];
 
         if ((DeviceCommand::WrapRegion)wrap == DeviceCommand::WrapRegion::COMPLETION) {
-            cq_write_interface.completion_fifo_wr_ptr = command_issue_region_size >> 4;     // Head to the beginning of the completion region
+            cq_write_interface.completion_fifo_wr_ptr = completion_queue_start_addr >> 4;     // Head to the beginning of the completion region
             cq_write_interface.completion_fifo_wr_toggle = not cq_write_interface.completion_fifo_wr_toggle;
+            // DPRINT << "Completion WRAPPING BACK TO " << (cq_write_interface.completion_fifo_wr_ptr << 4) << ENDL();
+            // for (volatile int i = 0; i < 10000000; i++);
```
Comment (abhullar-tt) [inline tt_metal/impl/dispatch/kernels/command_queue_consumer.cpp:63]: remove debug loop

**Review by abhullar-tt — COMMENTED** [review]
lgtm 🚀 ,  only thing I spoke to you about was the overloaded use of "channel" in the command queue apis. here I think we are referring to cq id but I can rename that in my changes that read from remote CQ (which is at a different channel)

File: tt_metal/core_descriptors/grayskull_120_arch.yaml [hunk 16]
```diff
@@ -17,8 +17,27 @@ E150:
...
+      start: [0, 0]
+      end: [11, 8]
+
+    storage_cores: # Relative to grid of tensix cores
+      [[2, -1],[3, -1],[4, -1],[5, -1],[8, -1],[9, -1],[10, -1],[11, -1]]
+
+    producer_cores:
+      [[0, -1], [1, -1]]
+
+    consumer_cores:
```
Comment (TT-billteng) [inline tt_metal/core_descriptors/grayskull_120_arch.yaml:39]: also, don't we need to support E75 for project grayskull? I see it still has "dispatch_cores"

File: tt_metal/host_api.hpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (TT-billteng) [reply tt_metal/host_api.hpp:63]: After re-reading your high-level PR description, we should try to address this in the follow-on PR

File: tt_metal/impl/dispatch/command_queue.hpp (hunk 10 above)
Comment (TT-billteng) [reply tt_metal/impl/dispatch/command_queue.hpp:194]: hope this will be addressed in the follow-on PR?

File: tt_metal/detail/tt_metal.hpp (hunk 9 above)
Comment (TT-billteng) [reply tt_metal/detail/tt_metal.hpp]: isn't this a bug?

File: tt_metal/hw/inc/mod_div_lib.h (hunk 3 above)
Comment (DrJessop) [reply tt_metal/hw/inc/mod_div_lib.h:50]: When we have multiple command queues, we need more cores, so we take away from the storage cores.  Given we need two storage cores, we remove 4 banks.

File: tt_metal/core_descriptors/grayskull_120_arch.yaml (hunk 16 above)
Comment (abhullar-tt) [reply tt_metal/core_descriptors/grayskull_120_arch.yaml:39]: also need to update WH yamls

File: tt_metal/impl/dispatch/command_queue.hpp (hunk 10 above)
Comment (DrJessop) [reply tt_metal/impl/dispatch/command_queue.hpp:194]: Good point. I didn't think about this. Yes, I can address in a follow-up.

File: tests/tt_metal/tt_metal/unit_tests_fast_dispatch/command_queue/test_EnqueueWriteBuffer_and_EnqueueReadBuffer.cpp (hunk 6 above)
Comment (DrJessop) [reply tests/tt_metal/tt_metal/unit_tests_fast_dispatch/command_queue/test_EnqueueWriteBuffer_and_EnqueueReadBuffer.cpp:330]: I don't think so.

File: tests/tt_metal/tt_metal/unit_tests_fast_dispatch/common/command_queue_fixture.hpp (hunk 7 above)
Comment (DrJessop) [reply tests/tt_metal/tt_metal/unit_tests_fast_dispatch/common/command_queue_fixture.hpp:38]: I think MultiCommandQueueFixture is gone now. And yes, good catch. Fixed.

File: tt_metal/detail/tt_metal.hpp (hunk 9 above)
Comment (DrJessop) [reply tt_metal/detail/tt_metal.hpp]: @TT-billteng  This is not a bug. Creating a command queue now is very light-weight (aka kernels not sent to device). I wanted to ensure that we don't ever get a case where the command queue is not defined for the device. I can add a comment.

File: tt_metal/detail/tt_metal.hpp (hunk 11 above)
Comment (DrJessop) [reply tt_metal/detail/tt_metal.hpp:384]: In one way, yes. The other problem is that I could not get @pgkeller 's dev message struct extended without getting hangs (needs to be extended in order for us to specify the dispatch core at runtime). So for now, I assert only one cq can be the program dispatcher.

File: tt_metal/detail/tt_metal.hpp (hunk 12 above)
Comment (DrJessop) [reply tt_metal/detail/tt_metal.hpp:455]: Yes

File: tt_metal/impl/dispatch/command_queue.cpp (hunk 13 above)
Comment (DrJessop) [reply tt_metal/impl/dispatch/command_queue.cpp:969]: No, I added it for debug. Can remove if you'd like.

File: tt_metal/impl/dispatch/kernels/command_queue_consumer.cpp (hunk 14 above)
Comment (DrJessop) [reply tt_metal/impl/dispatch/kernels/command_queue_consumer.cpp:74]: Got it.

File: tt_metal/impl/dispatch/kernels/command_queue_consumer.cpp (hunk 15 above)
Comment (DrJessop) [reply tt_metal/impl/dispatch/kernels/command_queue_consumer.cpp:63]: Got it.

File: tt_metal/core_descriptors/grayskull_120_arch.yaml (hunk 16 above)
Comment (DrJessop) [reply tt_metal/core_descriptors/grayskull_120_arch.yaml:39]: Fixed.

File: tt_metal/detail/tt_metal.hpp (hunk 11 above)
Comment (abhullar-tt) [reply tt_metal/detail/tt_metal.hpp:384]: ah makes sense, perhaps open an issue for that to track

**Review by TT-billteng — APPROVED** [review]

**Review by abhullar-tt — APPROVED** [review]

**Review by tt-aho — APPROVED** [review]
